File: striver_series/src/CountSubArrayGivenSumOptimize.java
import java.util.*;

public class CountSubArrayGivenSumOptimize {

    public static void main(String[] args) {
        int[] arr = {3, 1, 2, 4};
        int k = 6;
        int cnt = findAllSubarraysWithGivenSum(arr, k);
        System.out.println("The number of subarrays is: " + cnt);
    }

    private static int findAllSubarraysWithGivenSum(int[] arr, int k){
        int count = 0;
        Map<Integer,Integer> prefixSumCount = new HashMap<>();
        int sum = 0;
        prefixSumCount.put(0,1);
        for(int num :arr){
            sum+=num;
            int rem = sum -k;
            if(prefixSumCount.containsKey(rem)){
                count+= prefixSumCount.get(rem);
            }
            prefixSumCount.put(sum, prefixSumCount.getOrDefault(sum,0)+1);
        }
        return count;
    }
}

File: dart_programs/practise_programs/count_subarray_given_sum_optimize.dart

void main() {
  List<int> arr = [3, 1, 2, 4];
  int k = 6;
  int cnt = findAllSubarraysWithGivenSum(arr, k);
  print("The number of subarrays is: $cnt");
}

int findAllSubarraysWithGivenSum(List<int> arr, int k) {
  int count = 0;
  Map<int, int> prefixSumCount = {0: 1};
  int sum = 0;

  for (int num in arr) {
    sum += num;
    int rem = sum - k;
    if (prefixSumCount.containsKey(rem)) {
      count += prefixSumCount[rem]!;
    }
    prefixSumCount[sum] = (prefixSumCount[sum] ?? 0) + 1;
  }
  return count;
}
